File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/HomeCachePersistence.kt
package com.bockmedia.console.domain.model

import android.content.Context
import com.bockmedia.console.data.api.bockJson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import java.io.File

/**
 * Disk persistence for the home feed and its artwork lookups. Keeping a snapshot
 * on disk lets a cold launch paint the last-known home screen instantly while a
 * fresh copy loads in the background — instead of rebuilding from ~6 network
 * calls and re-resolving every cover every time.
 */
object HomeCachePersistence {
    private const val FILE_NAME = "home_cache.json"
    private const val MAX_AGE_MS = 24 * 60 * 60 * 1000L

    @Serializable
    private data class PlayTargetDto(
        val type: String,
        val id: String? = null,
        val name: String? = null,
        val artist: String? = null,
        val path: String? = null,
        val title: String? = null,
        val displayTitle: String? = null,
        val seedKind: String? = null,
    )

    @Serializable
    private data class CardDto(
        val id: String,
        val title: String,
        val subtitle: String? = null,
        val artPath: String? = null,
        val playlistId: String? = null,
        val kind: String,
        val playTarget: PlayTargetDto,
    )

    @Serializable
    private data class SectionDto(
        val id: String,
        val title: String,
        val kind: String,
        val cards: List<CardDto>,
    )

    @Serializable
    private data class SnapshotDto(
        val savedAtMs: Long,
        val sections: List<SectionDto>,
        val cardUrls: Map<String, String> = emptyMap(),
        val cardMediaPaths: Map<String, String> = emptyMap(),
        val playlistPaths: Map<String, String> = emptyMap(),
    )

    data class Snapshot(
        val feed: HomeFeed,
        val cardMediaPaths: Map<String, String>,
        val playlistPaths: Map<String, String>,
    )

    /** Serializes off the main thread and writes atomically (temp + rename). */
    suspend fun save(
        context: Context,
        feed: HomeFeed,
        cardMediaPaths: Map<String, String>,
        playlistPaths: Map<String, String>,
    ) {
        if (feed.sections.isEmpty()) return
        withContext(Dispatchers.IO) {
            runCatching {
                val dto = SnapshotDto(
                    savedAtMs = System.currentTimeMillis(),
                    sections = feed.sections.map { it.toDto() },
                    cardMediaPaths = cardMediaPaths,
                    playlistPaths = playlistPaths,
                )
                val json = bockJson.encodeToString(dto)
                val target = file(context)
                val tmp = File(target.parentFile, "$FILE_NAME.tmp")
                tmp.writeText(json)
                if (!tmp.renameTo(target)) {
                    target.writeText(json)
                    tmp.delete()
                }
            }
        }
    }

    suspend fun load(context: Context): Snapshot? = withContext(Dispatchers.IO) {
        runCatching {
            val f = file(context)
            if (!f.exists()) return@runCatching null
            val dto = bockJson.decodeFromString<SnapshotDto>(f.readText())
            if (System.currentTimeMillis() - dto.savedAtMs > MAX_AGE_MS) return@runCatching null
            val sections = dto.sections.mapNotNull { it.toModel() }
            if (sections.isEmpty()) return@runCatching null
            val cardPaths = dto.cardMediaPaths.ifEmpty {
                dto.cardUrls.mapNotNull { (id, url) ->
                    ArtworkPaths.extractMediaPath(url)?.let { id to it }
                }.toMap()
            }
            Snapshot(HomeFeed(sections), cardPaths, dto.playlistPaths)
        }.getOrNull()
    }

    fun clear(context: Context) {
        runCatching { file(context).delete() }
    }

    private fun file(context: Context) = File(context.filesDir, FILE_NAME)

    private fun HomeSection.toDto() = SectionDto(
        id = id,
        title = title,
        kind = kind.name,
        cards = cards.map { it.toDto() },
    )

    private fun HomeCard.toDto() = CardDto(
        id = id,
        title = title,
        subtitle = subtitle,
        artPath = artPath,
        playlistId = playlistId,
        kind = kind.name,
        playTarget = playTarget.toDto(),
    )

    private fun PlayTarget.toDto(): PlayTargetDto = when (this) {
        is PlayTarget.Playlist -> PlayTargetDto("playlist", id = id, name = name)
        is PlayTarget.Artist -> PlayTargetDto("artist", name = name)
        is PlayTarget.Album -> PlayTargetDto("album", name = name, artist = artist)
        is PlayTarget.Song -> PlayTargetDto("song", path = path, title = title)
        is PlayTarget.Radio -> PlayTargetDto(
            "radio",
            displayTitle = displayTitle,
            name = name,
            path = path,
            seedKind = seedKind.name,
        )
    }

    private fun SectionDto.toModel(): HomeSection? {
        val sectionKind = runCatching { HomeSectionKind.valueOf(kind) }.getOrNull() ?: return null
        val modelCards = cards.mapNotNull { it.toModel() }
        if (modelCards.isEmpty()) return null
        return HomeSection(id, title, sectionKind, modelCards)
    }

    private fun CardDto.toModel(): HomeCard? {
        val cardKind = runCatching { HomeSectionKind.valueOf(kind) }.getOrNull() ?: return null
        val target = playTarget.toModel() ?: return null
        return HomeCard(
            id = id,
            title = title,
            subtitle = subtitle,
            artPath = artPath,
            playlistId = playlistId,
            playTarget = target,
            kind = cardKind,
        )
    }

    private fun PlayTargetDto.toModel(): PlayTarget? = when (type) {
        "playlist" -> PlayTarget.Playlist(id.orEmpty(), name.orEmpty())
        "artist" -> name?.let { PlayTarget.Artist(it) }
        "album" -> name?.let { PlayTarget.Album(it, artist) }
        "song" -> path?.let { PlayTarget.Song(it, title ?: name ?: "Track") }
        "radio" -> {
            val seed = runCatching {
                PlayTarget.RadioSeedKind.valueOf(seedKind ?: "Artist")
            }.getOrDefault(PlayTarget.RadioSeedKind.Artist)
            PlayTarget.Radio(displayTitle ?: name.orEmpty(), seed, name.orEmpty(), path)
        }
        else -> null
    }
}

File: ios/BockMedia/Domain/HomeCachePersistence.swift
import Foundation

enum HomeCachePersistence {
    static let feedLayoutVersion = 3
    private static let fileName = "home_cache.json"
    private static let maxAgeMs: Int64 = 24 * 60 * 60 * 1000

    struct Snapshot {
        let feed: HomeFeed
        let playlistPaths: [String: String]
        let hasRatedSongs: Bool?
    }

    private struct SnapshotDto: Codable {
        let savedAtMs: Int64
        let feedVersion: Int
        let sections: [SectionDto]
        let playlistPaths: [String: String]?
        let hasRatedSongs: Bool?
    }

    private struct SectionDto: Codable {
        let id: String
        let title: String
        let kind: String
        let cards: [CardDto]
    }

    private struct CardDto: Codable {
        let id: String
        let title: String
        let subtitle: String?
        let artPath: String?
        let playlistId: String?
        let kind: String
        let playTarget: PlayTargetDto
    }

    private struct PlayTargetDto: Codable {
        let type: String
        let id: String?
        let name: String?
        let artist: String?
        let path: String?
        let title: String?
        let displayTitle: String?
        let seedKind: String?
    }

    static func save(_ feed: HomeFeed, playlistPaths: [String: String] = HomeArtworkCache.allPlaylistPaths()) {
        guard !feed.sections.isEmpty else { return }
        let dto = SnapshotDto(
            savedAtMs: Int64(Date().timeIntervalSince1970 * 1000),
            feedVersion: feedLayoutVersion,
            sections: feed.sections.map(sectionToDto),
            playlistPaths: playlistPaths.isEmpty ? nil : playlistPaths,
            hasRatedSongs: HomeFeedCache.peekHasRatedSongs()
        )
        guard let url = cacheURL(), let json = try? JSONEncoder().encode(dto) else { return }
        try? json.write(to: url, options: .atomic)
    }

    static func load() -> Snapshot? {
        guard let url = cacheURL(),
              let data = try? Data(contentsOf: url),
              let dto = try? JSONDecoder().decode(SnapshotDto.self, from: data) else { return nil }
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        if now - dto.savedAtMs > maxAgeMs || dto.feedVersion < feedLayoutVersion { return nil }
        let sections = dto.sections.compactMap(dtoToSection)
        guard !sections.isEmpty else { return nil }
        let feed = HomeFeed(sections: sections)
        guard feed.isUsableHomeCache(
            activeProfileLinked: ActiveProfileStore.activeMemberId() != nil,
            hasRatedSongs: dto.hasRatedSongs
        ) else { return nil }
        return Snapshot(feed: feed, playlistPaths: dto.playlistPaths ?? [:], hasRatedSongs: dto.hasRatedSongs)
    }

    static func clear() {
        guard let url = cacheURL() else { return }
        try? FileManager.default.removeItem(at: url)
    }

    private static func cacheURL() -> URL? {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first?
            .appendingPathComponent(fileName)
    }

    private static func sectionToDto(_ section: HomeSection) -> SectionDto {
        SectionDto(
            id: section.id,
            title: section.title,
            kind: kindName(section.kind),
            cards: section.cards.map(cardToDto)
        )
    }

    private static func dtoToSection(_ dto: SectionDto) -> HomeSection? {
        guard let kind = parseKind(dto.kind) else { return nil }
        let cards = dto.cards.compactMap(dtoToCard)
        guard !cards.isEmpty else { return nil }
        return HomeSection(id: dto.id, title: dto.title, kind: kind, cards: cards)
    }

    private static func cardToDto(_ card: HomeCard) -> CardDto {
        CardDto(
            id: card.id,
            title: card.title,
            subtitle: card.subtitle,
            artPath: card.artPath,
            playlistId: card.playlistId,
            kind: kindName(card.kind),
            playTarget: targetToDto(card.playTarget)
        )
    }

    private static func dtoToCard(_ dto: CardDto) -> HomeCard? {
        guard let kind = parseKind(dto.kind), let target = dtoToTarget(dto.playTarget) else { return nil }
        return HomeCard(
            id: dto.id,
            title: dto.title,
            subtitle: dto.subtitle,
            artPath: dto.artPath,
            playlistId: dto.playlistId,
            playTarget: target,
            kind: kind
        )
    }

    private static func kindName(_ kind: HomeSectionKind) -> String {
        switch kind {
        case .jumpBackIn: return "jumpBackIn"
        case .favorites: return "favorites"
        case .ratedSongs: return "ratedSongs"
        case .topMixes: return "topMixes"
        case .browseGenres: return "browseGenres"
        case .exploreThemes: return "exploreThemes"
        case .mood: return "mood"
        case .decade: return "decade"
        case .dailyMixes: return "dailyMixes"
        case .recentlyCreated: return "recentlyCreated"
        case .recentPlaylists: return "recentPlaylists"
        case .radio: return "radio"
        case .discover: return "discover"
        case .offline: return "offline"
        }
    }

    private static func parseKind(_ raw: String) -> HomeSectionKind? {
        switch raw {
        case "jumpBackIn": return .jumpBackIn
        case "favorites": return .favorites
        case "ratedSongs": return .ratedSongs
        case "topMixes": return .topMixes
        case "browseGenres": return .browseGenres
        case "exploreThemes": return .exploreThemes
        case "mood": return .mood
        case "decade": return .decade
        case "dailyMixes": return .dailyMixes
        case "recentlyCreated": return .recentlyCreated
        case "recentPlaylists": return .recentPlaylists
        case "radio": return .radio
        case "discover": return .discover
        case "offline": return .offline
        default: return nil
        }
    }

    private static func targetToDto(_ target: PlayTarget) -> PlayTargetDto {
        switch target {
        case .playlist(let id, let name):
            return PlayTargetDto(type: "playlist", id: id, name: name, artist: nil, path: nil, title: nil, displayTitle: nil, seedKind: nil)
        case .artist(let name):
            return PlayTargetDto(type: "artist", id: nil, name: name, artist: nil, path: nil, title: nil, displayTitle: nil, seedKind: nil)
        case .album(let name, let artist):
            return PlayTargetDto(type: "album", id: nil, name: name, artist: artist, path: nil, title: nil, displayTitle: nil, seedKind: nil)
        case .song(let path, let title):
            return PlayTargetDto(type: "song", id: nil, name: nil, artist: nil, path: path, title: title, displayTitle: nil, seedKind: nil)
        case .radio(let displayTitle, let seedKind, let name, let path):
            let seed = switch seedKind {
            case .artist: "Artist"
            case .song: "Song"
            case .genre: "Genre"
            }
            return PlayTargetDto(type: "radio", id: nil, name: name, artist: nil, path: path, title: nil, displayTitle: displayTitle, seedKind: seed)
        }
    }

    private static func dtoToTarget(_ dto: PlayTargetDto) -> PlayTarget? {
        switch dto.type {
        case "playlist":
            return .playlist(id: dto.id ?? "", name: dto.name ?? "")
        case "artist":
            guard let name = dto.name else { return nil }
            return .artist(name: name)
        case "album":
            guard let name = dto.name else { return nil }
            return .album(name: name, artist: dto.artist)
        case "song":
            guard let path = dto.path else { return nil }
            return .song(path: path, title: dto.title ?? dto.name ?? "Track")
        case "radio":
            let seed: PlayTarget.RadioSeedKind = switch dto.seedKind?.lowercased() {
            case "song": .song
            case "genre": .genre
            default: .artist
            }
            return .radio(displayTitle: dto.displayTitle ?? dto.name ?? "", seedKind: seed, name: dto.name ?? "", path: dto.path)
        default:
            return nil
        }
    }
}

// hasCurrentHomeLayout / isUsableHomeCache live on HomeFeed in HomeFeed.swift
